src/client.js

```
@@ -0,0 +1,442 @@
+    } else {
+      // Use default global agent(s)...
+      this._httpAgent = isHttps ?
+        DEFAULT_AGENTS.https :
+        DEFAULT_AGENTS.http;
+    }
+
+    // Timeout for each _individual_ http request.
+    this._timeout = this._options.timeout;
+
+    const { credentials } = this._options;
+
+    if (credentials && credentials.clientId && credentials.accessToken) {
+      // Build ext for hawk requests
+      const ext = {};
+
+      if (credentials.certificate) {
+        // If there is a certificate we have temporary credentials, and we
+        // must provide the certificate
+        if (typeof credentials.certificate === 'string') {
+          // Parse as JSON if it's a string
+          try {
+            ext.certificate = JSON.parse(credentials.certificate);
+          } catch (err) {
+            debug('Failed to parse credentials.certificate, err: %s, JSON: %j', err, err);
+            throw new Error('Failed to parse configured certificate as valid JSON');
+          }
+        } else {
+          ext.certificate = credentials.certificate;
+        }
+      }
+
+      // If set of authorized scopes is provided, we'll restrict the request
+      // to only use these scopes
+      if (Array.isArray(this._options.authorizedScopes)) {
+        ext.authorizedScopes = this._options.authorizedScopes;
+      }
+
+      // If ext has any keys we base64 encode it and set ext on extra
+      if (Object.keys(ext).length) {
+        this._extData = new Buffer(JSON.stringify(ext)).toString('base64');
+      }
+    }
+  }
+
+  /** Make a request for a Client instance */
+  async makeRequest(entry, url, payload, query) {
+    const { randomizationFactor: rf, delayFactor, retries, maxDelay } = this._options;
+
+    // Add query to url if present
+    if (query) {
+      query = querystring.stringify(query);
+
+      if (query.length) {
+        url += `?${query}`;
+      }
+    }
+
+    // Construct request object
+    const options = {
+      json: true,
```

I'm not sure what this does... but we want to throw an error with a statusCode property, if the response isn't json.